src/kafka/connection/tcp_connection.hh

```diff
@@ -0,0 +1,69 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Copyright (C) 2019 ScyllaDB Ltd.
+ */
+
+#pragma once
+
+#include <seastar/core/future.hh>
+#include <seastar/net/net.hh>
+#include <seastar/net/inet_address.hh>
+#include <string>
+
+namespace seastar {
+
+namespace kafka {
+
+struct tcp_connection_exception : public std::runtime_error {
+public:
+    explicit tcp_connection_exception(const std::string& message) : runtime_error(message) {}
+};
+
+class tcp_connection {
+
+    net::inet_address _host;
+    uint16_t _port;
+    uint32_t _timeout_ms;
+    connected_socket _fd;
+    input_stream<char> _read_buf;
+    output_stream<char> _write_buf;
+
+public:
+    static future<lw_shared_ptr<tcp_connection>> connect(const std::string& host, uint16_t port, uint32_t timeout_ms);
+
+    tcp_connection(const net::inet_address& host, uint16_t port, uint32_t timeout_ms, connected_socket&& fd) noexcept
```

Comment:
Can this be private if we have a static factory method connect?

Reply:
Unfortunately I don't remember where exactly as this was very long ago, but I did spend a fair amount of time of trying to make it private and there were problems. If I remember correctly it prevented shared ptrs to be constructed containing this class.

Reply:
This should be doable but since we don't need shared_ptr here it will be no problem to make the constructor private.

Reply:
I changed it to use raw tcp_connection objects and I get this https://pastebin.com/t9jrceyQ and to be honest, I don't really get what this error might mean. Still digging around, but would love some help if you've seen this before

Reply:
It seems that something that contains tcp_connection is being copied and tcp_connection has no copy constructor/assignment.

Reply:
It's something deeper though. Is the tuple mentioned in the error possibly something created inside a future? Are futures copying the values passed to callbacks at any given point?